gacl/admin/securityPatterns.php

```diff
@@ -0,0 +1,36 @@
+<?php
+function applyIVPattern($data)
+{
+$data = trim($data);
+$data = htmlspecialchars($data);
+
+// Fix escape sequqnce &entity\n;
+$data = str_replace(array('&amp;','&lt;','&gt;'), array('&amp;amp;','&amp;lt;','&amp;gt;'), $data);
+$data = preg_replace('/(&#*\w+)[\x00-\x20]+;/u', '$1;', $data);
+$data = preg_replace('/(&#x*[0-9A-F]+);*/iu', '$1;', $data);
+$data = html_entity_decode($data, ENT_COMPAT, 'UTF-8');
+
+// Remove any attribute starting potentiaon with "on" or xmlns
+$data = preg_replace('#(<[^>]+?[\x00-\x20"\'])(?:on|xmlns)[^>]*+>#iu', '$1>', $data);
+
+// Remove javascript: and vbscript: protocols
+$data = preg_replace('#([a-z]*)[\x00-\x20]*=[\x00-\x20]*([`\'"]*)[\x00-\x20]*j[\x00-\x20]*a[\x00-\x20]*v[\x00-\x20]*a[\x00-\x20]*s[\x00-\x20]*c[\x00-\x20]*r[\x00-\x20]*i[\x00-\x20]*p[\x00-\x20]*t[\x00-\x20]*:#iu', '$1=$2nojavascript...', $data);
+$data = preg_replace('#([a-z]*)[\x00-\x20]*=([\'"]*)[\x00-\x20]*v[\x00-\x20]*b[\x00-\x20]*s[\x00-\x20]*c[\x00-\x20]*r[\x00-\x20]*i[\x00-\x20]*p[\x00-\x20]*t[\x00-\x20]*:#iu', '$1=$2novbscript...', $data);
+$data = preg_replace('#([a-z]*)[\x00-\x20]*=([\'"]*)[\x00-\x20]*-moz-binding[\x00-\x20]*:#u', '$1=$2nomozbinding...', $data);
+
+// Remove namespace elements 
+$data = preg_replace('#</*\w+:\w[^>]*+>#i', '', $data);
+
+do
+{
+    // Remove unwanted tags
+    $old_data = $data;
+    $data = preg_replace('#</*(?:applet|b(?:ase|gsound|link)|embed|frame(?:set)?|i(?:frame|layer)|l(?:ayer|ink)|meta|object|s(?:cript|tyle)|title|xml)[^>]*+>#i', '', $data);
+}
+while ($old_data !== $data);
+
+// we can add more sanitize or escape sequence in the future (as the need arises)
+
+return $data;
```

Comment:
Wow, there is a lot going on here. I was under the impression that only the htmlspecialchars() call was needed to escape text for html. Is all the other processing really necessary for html escaping strings?